File: Ver/IA-Advanced-Search.py
import requests
from requests.adapters import HTTPAdapter
from urllib3.util.retry import Retry
import time
import json

SEARCH_URL = "https://archive.org/advancedsearch.php"
METADATA_BASE_URL = "https://archive.org/metadata/"
DOWNLOAD_BASE_URL = "https://archive.org/download"

# Build a valid query:
# - Only software media type
# - ISO or IMG files are likely
# - Mention "linux" in title/subject/description (subject:"linux" is common)
QUERY = '(format:ISO OR format:IMG) AND mediatype:software AND description:"linux, distribution"'

FIELDS = ["identifier", "title", "date", "creator"]
ROWS_PER_PAGE = 500  # IA allows up to 1000; use 500 to be gentle
REQUEST_TIMEOUT = 30
SLEEP_SECONDS = 1.0  # rate limiting between requests

# Configure a resilient HTTP session with retries and backoff
_SESSION = requests.Session()
_SESSION.headers.update({
    "User-Agent": "Internet-Archive-API/1.0 (+https://example.local) Python-requests"
})
_retry = Retry(
    total=5,
    connect=5,
    read=5,
    backoff_factor=1.0,
    status_forcelist=(429, 500, 502, 503, 504),
    allowed_methods=("HEAD", "GET", "OPTIONS"),
    raise_on_status=False,
)
_adapter = HTTPAdapter(max_retries=_retry)
_SESSION.mount("https://", _adapter)
_SESSION.mount("http://", _adapter)

def search_page(page: int) -> dict:
    params = {
        "q": QUERY,
        "fl[]": FIELDS,
        "rows": ROWS_PER_PAGE,
        "page": page,
        "output": "json"
    }
    try:
        resp = _SESSION.get(SEARCH_URL, params=params, timeout=REQUEST_TIMEOUT)
    except requests.RequestException as e:
        raise RuntimeError(f"Advanced search request error on page {page}: {e}") from e
    if resp.status_code != 200:
        raise RuntimeError(f"Advanced search failed with status {resp.status_code}: {resp.text[:300]}")
    try:
        data = resp.json()
    except json.JSONDecodeError as e:
        raise RuntimeError(f"Failed to parse JSON from advanced search: {e}\nBody: {resp.text[:300]}")
    return data

def fetch_metadata(identifier: str) -> dict | None:
    url = f"{METADATA_BASE_URL}{identifier}"
    try:
        resp = _SESSION.get(url, timeout=REQUEST_TIMEOUT)
    except requests.RequestException:
        return None
    if resp.status_code != 200:
        return None
    try:
        return resp.json()
    except json.JSONDecodeError:
        return None

def main():
    iso_entries = []

    # Fetch first page to get numFound and validate structure
    first = search_page(1)
    response_obj = first.get("response")
    if not isinstance(response_obj, dict) or "docs" not in response_obj:
        # Provide more context for debugging
        err = first.get("error") or first
        raise RuntimeError(f"Unexpected search response structure, missing 'response.docs'. Details: {json.dumps(err)[:500]}")

    num_found = int(response_obj.get("numFound", 0))
    pages = max(1, (num_found + ROWS_PER_PAGE - 1) // ROWS_PER_PAGE)

    for page in range(1, pages + 1):
        if page > 1:
            time.sleep(SLEEP_SECONDS)
            data = search_page(page)
            response_obj = data.get("response", {})
        docs = response_obj.get("docs", [])
        if not isinstance(docs, list):
            continue

        for item in docs:
            identifier = item.get("identifier")
            if not identifier:
                continue

            time.sleep(SLEEP_SECONDS)
            meta_json = fetch_metadata(identifier)
            if not meta_json:
                continue

            files = meta_json.get("files", []) or []
            for f in files:
                name = f.get("name", "") or ""
                lname = name.lower()
                if lname.endswith((".iso", ".img", ".zip")):
                    iso_entries.append({
                        "identifier": identifier,
                        "title": item.get("title", ""),
                        "file_name": name,
                        "download_url": f"{DOWNLOAD_BASE_URL}/{identifier}/{name}",
                        "size": f.get("size", "unknown")
                    })

    out_file = "../iso_metadata.json"
    with open(out_file, "w", encoding="utf-8") as f:
        json.dump(iso_entries, f, indent=2, ensure_ascii=False)

    print(f"Found {len(iso_entries)} ISO-like files. Saved to {out_file}.")

if __name__ == "__main__":
    main()

File: rust/src/bin/ia-advanced-search.rs
use std::{thread, time::Duration, fs::File, io::Write};

use anyhow::{Context, Result};
use clap::Parser;
use reqwest::{Client, StatusCode};
use serde::{Deserialize, Serialize};
use serde_json::Value;

const SEARCH_URL: &str = "https://archive.org/advancedsearch.php";
const METADATA_BASE_URL: &str = "https://archive.org/metadata/";
const DOWNLOAD_BASE_URL: &str = "https://archive.org/download";

#[derive(Parser, Debug)]
#[command(name = "ia-advanced-search", about = "Internet Archive Advanced Search (Rust)")]
struct Args {
    /// Advanced search query string
    #[arg(short, long, default_value = "(format:ISO OR format:IMG) AND mediatype:software AND description:\"linux, distribution\"")]
    query: String,
    /// Rows per page (<=1000)
    #[arg(long, default_value_t = 500)]
    rows: usize,
    /// Limit number of pages to fetch
    #[arg(long)]
    max_pages: Option<usize>,
    /// Sleep seconds between requests
    #[arg(long, default_value_t = 1.0)]
    sleep: f32,
    /// Fields to fetch in search results
    #[arg(long, num_args = 1.., value_delimiter = ' ', default_values_t = ["identifier".to_string(), "title".to_string(), "date".to_string(), "creator".to_string()])]
    fields: Vec<String>,
    /// Output JSON file for results
    #[arg(short, long, default_value = "iso_metadata.json")]
    out: String,
    /// Request timeout seconds
    #[arg(long, default_value_t = 30)]
    timeout: u64,
    /// HTTP retries for transient errors
    #[arg(long, default_value_t = 5)]
    retries: usize,
    /// Retry backoff factor
    #[arg(long, default_value_t = 1.0)]
    backoff: f32,
    /// Custom User-Agent header
    #[arg(long)]
    user_agent: Option<String>,
    /// Increase verbosity (-v info, -vv debug)
    #[arg(short = 'v', action = clap::ArgAction::Count)]
    verbosity: u8,
    /// Do not fetch per-item metadata, only list identifiers
    #[arg(long)]
    dry_run: bool,
}

#[derive(Debug, Deserialize)]
struct SearchResponse {
    response: Option<SearchInner>,
    #[allow(dead_code)]
    error: Option<Value>,
}

#[derive(Debug, Deserialize)]
struct SearchInner {
    #[serde(default, rename = "numFound")]
    num_found: i64,
    #[serde(default)]
    docs: Vec<serde_json::Map<String, Value>>, // flexible
}

#[derive(Debug, Serialize, Deserialize)]
struct IsoEntry {
    identifier: String,
    title: String,
    file_name: String,
    download_url: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    size: Option<String>,
}

#[tokio::main(flavor = "multi_thread")]
async fn main() -> Result<()> {
    let args = Args::parse();

    let client = Client::builder()
        .user_agent(args.user_agent.clone().unwrap_or_else(|| "Internet-Archive-API/rs".to_string()))
        .timeout(Duration::from_secs(args.timeout))
        .build()?;

    if args.verbosity >= 1 {
        eprintln!("Query: {}", args.query);
    }

    let mut iso_entries: Vec<IsoEntry> = Vec::new();

    // First page
    let first: SearchResponse = get_with_retries_json(&client, SEARCH_URL, args.retries, args.backoff, |u| {
        {
            let mut qp = u.query_pairs_mut();
            qp.append_pair("q", args.query.as_str());
            qp.append_pair("rows", &args.rows.to_string());
            qp.append_pair("page", "1");
            qp.append_pair("output", "json");
            for f in &args.fields {
                qp.append_pair("fl[]", f.as_str());
            }
        }
    }).await?;

    let mut resp_obj = first.response.context("Unexpected search response structure, missing 'response'")?;
    let num_found = resp_obj.num_found.max(0) as usize;
    let mut total_pages = ((num_found + args.rows - 1).max(1)) / args.rows;
    if num_found > 0 && num_found % args.rows != 0 { total_pages += 1; }
    if let Some(maxp) = args.max_pages { total_pages = total_pages.min(maxp); }

    if args.verbosity >= 1 {
        eprintln!("numFound={}, pages={}", num_found, total_pages);
    }

    for page in 1..=total_pages {
        if page > 1 {
            thread::sleep(Duration::from_secs_f32(args.sleep));
            let data: SearchResponse = get_with_retries_json(&client, SEARCH_URL, args.retries, args.backoff, |u| {
                {
                    let mut qp = u.query_pairs_mut();
                    qp.append_pair("q", args.query.as_str());
                    qp.append_pair("rows", &args.rows.to_string());
                    qp.append_pair("page", &page.to_string());
                    qp.append_pair("output", "json");
                    for f in &args.fields { qp.append_pair("fl[]", f.as_str()); }
                }
            }).await?;
            if let Some(inner) = data.response { resp_obj = inner; }
        }

        let docs = &resp_obj.docs;
        if args.verbosity >= 2 {
            eprintln!("Processing page {} with {} docs", page, docs.len());
        }

        for item in docs {
            let identifier = item.get("identifier").and_then(|v| v.as_str()).unwrap_or("");
            if identifier.is_empty() { continue; }
            let title = item.get("title").and_then(|v| v.as_str()).unwrap_or("").to_string();

            if args.dry_run {
                println!("{} - {}", identifier, title);
                continue;
            }

            thread::sleep(Duration::from_secs_f32(args.sleep));
            let meta_url = format!("{}{}", METADATA_BASE_URL, identifier);
            let meta: Option<Value> = match get_with_retries_json_opt(&client, &meta_url, args.retries, args.backoff).await {
                Ok(v) => v,
                Err(_) => None,
            };
            if let Some(Value::Object(map)) = meta {
                if let Some(files) = map.get("files").and_then(|v| v.as_array()) {
                    for f in files {
                        if let Some(name) = f.get("name").and_then(|v| v.as_str()) {
                            let lname = name.to_lowercase();
                            if lname.ends_with(".iso") || lname.ends_with(".img") || lname.ends_with(".zip") {
                                let size = f.get("size").and_then(|v| v.as_i64()).map(|n| n.to_string());
                                iso_entries.push(IsoEntry {
                                    identifier: identifier.to_string(),
                                    title: title.clone(),
                                    file_name: name.to_string(),
                                    download_url: format!("{}/{}/{}", DOWNLOAD_BASE_URL, identifier, name),
                                    size,
                                });
                            }
                        }
                    }
                }
            }
        }
    }

    let mut file = File::create(&args.out).with_context(|| format!("Failed to create {}", &args.out))?;
    file.write_all(serde_json::to_string_pretty(&iso_entries)?.as_bytes())?;
    println!("Found {} ISO-like files. Saved to {}.", iso_entries.len(), &args.out);

    Ok(())
}

async fn get_with_retries_json<T: for<'de> serde::Deserialize<'de>, F: FnOnce(&mut reqwest::Url)>(client: &Client, base: &str, retries: usize, backoff: f32, url_mut: F) -> Result<T> {
    let mut url = reqwest::Url::parse(base)?;
    url_mut(&mut url);
    let mut attempt = 0usize;
    loop {
        attempt += 1;
        let res = client.get(url.clone()).send().await;
        match res {
            Ok(resp) => {
                if resp.status() == StatusCode::OK {
                    let v = resp.json::<T>().await?;
                    return Ok(v);
                } else if matches!(resp.status(), StatusCode::TOO_MANY_REQUESTS | StatusCode::INTERNAL_SERVER_ERROR | StatusCode::BAD_GATEWAY | StatusCode::SERVICE_UNAVAILABLE | StatusCode::GATEWAY_TIMEOUT) {
                    // retryable
                } else {
                    let status = resp.status();
                    let text = resp.text().await.unwrap_or_default();
                    anyhow::bail!("Request failed with status {}: {}", status, text);
                }
            }
            Err(_) => { /* retry */ }
        }
        if attempt > retries { anyhow::bail!("Failed after {} retries", retries); }
        let sleep = backoff * attempt as f32;
        thread::sleep(Duration::from_secs_f32(sleep));
    }
}

async fn get_with_retries_json_opt(client: &Client, url: &str, retries: usize, backoff: f32) -> Result<Option<Value>> {
    let mut attempt = 0usize;
    loop {
        attempt += 1;
        match client.get(url).send().await {
            Ok(resp) => {
                if resp.status().is_success() {
                    let v: Value = resp.json().await?;
                    return Ok(Some(v));
                } else if matches!(resp.status(), StatusCode::TOO_MANY_REQUESTS | StatusCode::INTERNAL_SERVER_ERROR | StatusCode::BAD_GATEWAY | StatusCode::SERVICE_UNAVAILABLE | StatusCode::GATEWAY_TIMEOUT) {
                    // retryable
                } else {
                    return Ok(None);
                }
            }
            Err(_) => { /* retry */ }
        }
        if attempt > retries { return Ok(None); }
        let sleep = backoff * attempt as f32;
        thread::sleep(Duration::from_secs_f32(sleep));
    }
}
